internal/pkg/action/operator_uninstall.go

```diff
@@ -165,28 +152,72 @@ func (u *OperatorUninstall) getSubscriptionCSV(ctx context.Context, subscription
 	return csv, name, nil
 }
 
-func csvNameFromSubscription(subscription *v1alpha1.Subscription) string {
-	if subscription.Status.InstalledCSV != "" {
-		return subscription.Status.InstalledCSV
+func (u *OperatorUninstall) deleteOperatorGroup(ctx context.Context) error {
+	subs := v1alpha1.SubscriptionList{}
+	if err := u.config.Client.List(ctx, &subs, client.InNamespace(u.config.Namespace)); err != nil {
+		return fmt.Errorf("list subscriptions: %v", err)
 	}
-	return subscription.Status.CurrentCSV
+
+	// If there are no subscriptions left, delete the operator group(s).
+	if len(subs.Items) == 0 {
+		ogs := v1.OperatorGroupList{}
+		if err := u.config.Client.List(ctx, &ogs, client.InNamespace(u.config.Namespace)); err != nil {
+			return fmt.Errorf("list operatorgroups: %v", err)
+		}
+		for _, og := range ogs.Items {
+			og := og
+			if len(u.DeleteOperatorGroupNames) == 0 || contains(u.DeleteOperatorGroupNames, og.GetName()) {
+				if err := u.deleteObjects(ctx, &og); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
 }
 
-// getCRDs returns the list of CRDs required by a CSV.
-func getCRDs(csv *v1alpha1.ClusterServiceVersion) (crds []client.Object) {
-	for _, resource := range csv.Status.RequirementStatus {
-		if resource.Kind == crdKind {
-			obj := &unstructured.Unstructured{}
-			obj.SetGroupVersionKind(schema.GroupVersionKind{
-				Group:   resource.Group,
-				Version: resource.Version,
-				Kind:    resource.Kind,
-			})
-			obj.SetName(resource.Name)
-			crds = append(crds, obj)
+// validStrategy validates the deletion strategy against the operands on-cluster
+func (u *OperatorUninstall) validStrategy(operands *unstructured.UnstructuredList) (bool, error) {
+	if len(operands.Items) > 0 && u.OperandStrategy.Kind == operand.Cancel {
+		return false, fmt.Errorf("%d operands exist and operand strategy %q is in use: "+
+			"delete operands manually or re-run uninstall with a different operand deletion strategy."+
+			"\n\nSee kubectl operator uninstall --help for more information on operand deletion strategies.", len(operands.Items), operand.Cancel)
```

Comment:
This is fine for now, but I think we should follow up and rework the way this error message works. Theoretically, once we export the `OperatorUninstall` struct in a non-internal tree, this function could be imported and used by other tools (e.g. operator-sdk's `cleanup` subcommand), so it would be better not reference the CLI in this package.

I think we could rework this by defining an error struct e.g. `OperandStrategyError` that the cmd layer can use `errors.As` on, and we could then put the CLI-related log messages in the cmd layer instead of the `action` package.

Could you add a TODO comment as a reminder to ourselves?

Reply:
Makes total sense